File: src/main/java/ThirtyDayLeetCodingChallenge/Day23_BitwiseAndOfNumbersRange.java
package ThirtyDayLeetCodingChallenge;

public class Day23_BitwiseAndOfNumbersRange {
    public static void main(String[] args) {
        System.out.println(rangeBitwiseAnd(5, 7));
        System.out.println(rangeBitwiseAnd(0, 1));
        System.out.println(rangeBitwiseAnd(2147483646, 2147483647));
    }

    public static int rangeBitwiseAnd(int m, int n) {
        if (m == n) return m;

        int result = m;

        for (long i = m + 1; i <= n; i++) {
            result = result & (int) i;
            if (result == 0) return 0;
        }

        return result;
    }
}

File: src/main/kotlin/April2020/Day23_BitwiseAndOfNumbersRange.kt
fun main() {
    println(rangeBitwiseAnd(5, 7))
    println(rangeBitwiseAnd(0, 1))
    println(rangeBitwiseAnd(2147483646, 2147483647))
    println(rangeBitwiseAnd(2147483647, 2147483647))
}

fun rangeBitwiseAnd(m: Int, n: Int): Int =
    if (n > m) (rangeBitwiseAnd(m/2, n/2) shl 1) else m
